### ISSUE TITLE
cast cast cast cast cast abstract inherited strange

### ISSUE THREAD

**Issue by nanjizal** [HaxeFoundation/haxe#9451]
**cast cast cast cast cast abstract inherited strange**
using the ```-D dump=pretty```
I get multiple 'casts' it the optimised dump?
https://gist.github.com/nanjizal/abd3bf8b7a0441d353ff9ef2146c305d
currently the class looks like but it could change.
https://github.com/nanjizal/dice/blob/master/src/dice/Dice.hx
 don't have full unit tests coverage in the **geom** library so the full multiple cast are not used in tests or dox, but I checked in with dump so you can see it starting to occur.
https://github.com/nanjizal/geom/blob/master/dump/js/geom/flat/f32/_Float32Flat/Float32Flat_Impl_.dump#L70
Basically the code is just a triple abstract wrapper around Float32Array, partly to avoid problems with not having self and limitations on array access, it stores position and length at [0],[1] it provides a view at each position so you can easily set triangle positions or RGBA values. I believe casts are being accumulated at each abstraction, I am not even sure they are needed.

On a side note abstract.abstract_impl seem kind of rather long could this be less verbose?

**Comment by RealyUniqueName** [comment]
Those casts usually pop up during various code transformations (e.g. multiple inlined methods or abstract casts) and are required to not lose intermediate types. 
In general multiple sequential casts get removed by target generators and don't impose unjustified overhead.

**Comment by nanjizal** [comment]
ok it seems to go away in js. Just weird that they are in the dump, but I get that it might be safer not to try to remove them, just a bit scary when you see them in the dump.
Probably better if they were not repeated for any user evaluating haxe as it is rather confusing, but I guess maybe they are less likely to be abusing abstracts as much :)

**Comment by nadako** [comment]
For C#/Java there's a gencommon filter that tries to strip redundant casts before they reach generators   (which will generate actual code for them) and IMO it would be useful to have this is in a unified and configurable way, not unlike the new var renaming filter.

**Comment by jdonaldson** [comment]
FWIW I ran into a similar situation with Std.string.  In some cases you will see things like Std.string(Std.String(...)).

I was wondering at the time whether it would be useful to mark some functions as idempotent, at least with regards to certain type parameters.
